src/systemcmds/bsondump/bsondump.cpp

```diff
@@ -82,58 +91,224 @@ static int bson_print_callback(bson_decoder_t decoder, bson_node_t node)
 	return -1;
 }
 
+static int
+bson_docsize_callback(bson_decoder_t decoder, bson_node_t node)
+{
+
+	if (node->type == BSON_EOO) {
+		PX4_DEBUG("end of parameters");
+		return 0;
+	}
+
+	return 1;
+}
+
+int copy_file(const char *src_path, const char *dst_path)
+{
+	int src_fd = open(src_path, O_RDONLY);
+	int dst_fd = open(dst_path, O_WRONLY | O_TRUNC);
+
+	if (src_fd == -1 || dst_fd == -1) {
+		perror("Failed to open files for copying");
+
+		if (src_fd != -1) { close(src_fd); }
+
+		if (dst_fd != -1) { close(dst_fd); }
+
+		return -1;
+	}
+
+	char buffer[1024];
+	ssize_t bytes_read;
+
+	while ((bytes_read = read(src_fd, buffer, sizeof(buffer))) > 0) {
+		if (write(dst_fd, buffer, bytes_read) != bytes_read) {
+			perror("Failed to copy file");
+			close(src_fd);
+			close(dst_fd);
+			return -1;
+		}
+	}
+
+	close(src_fd);
+	close(dst_fd);
+
+	if (bytes_read == -1) {
+		perror("Error reading from source file");
+		return -1;
+	}
+
+	return 0;
+}
+
 extern "C" __EXPORT int bsondump_main(int argc, char *argv[])
 {
-	if (argc != 2) {
-		print_usage();
+	if (argc < 2) {
+		print_usage("Invalid number of arguments.");
+		return -1;
+	}
+
+	char *command = argv[1];
+	char *file_name = argv[3];
+
+	if (strcmp(command, "docsize") == 0) {
+		if (argc != 3) {
+			print_usage("Usage: bsondump docsize <file>", "docsize");
+			return -1;
+		}
+
+		file_name = argv[2];
+		int source_fd = open(file_name, O_RDONLY);
+
+		if (source_fd < 0) {
+			PX4_ERR("open '%s' failed (%i)", file_name, errno);
+			return 1;
+		}
+
+		bson_decoder_s decoder{};
+		int result = -1;
+
+		if (bson_decoder_init_file(&decoder, source_fd, bson_docsize_callback) == 0) {
+
+			do {
+				result = bson_decoder_next(&decoder);
+
+			} while (result > 0);
+
+			PX4_INFO("DECODED_SIZE:%" PRId32 " SAVED_SIZE:%" PRId32 "\n",
+				 decoder.total_decoded_size, decoder.total_document_size);
+		}
+
+		close(source_fd);
+
+		if (decoder.total_decoded_size != decoder.total_document_size && decoder.total_document_size == 0
+		    && strstr(file_name, "mtd_caldata") != NULL) {
```

Comment:
```suggestion
) {
```

Reply:
The assumption was that this is only an issue for mtd_caldata so the fix would only apply if the file_path included mtd_caldata. Should we assume that it can be any bson file?

In that case, I would append the paths with the last name of file_name.